docker/docker_client.go

```diff
@@ -592,8 +642,10 @@ func (c *dockerClient) setupRequestAuth(req *http.Request, extraScope *authScope
 				cacheKey := ""
 				scopes := []authScope{c.scope}
 				if extraScope != nil {
-					// Using ':' as a separator here is unambiguous because getBearerToken below uses the same separator when formatting a remote request (and because repository names can't contain colons).
-					cacheKey = fmt.Sprintf("%s:%s", extraScope.remoteName, extraScope.actions)
+					// Using ':' as a separator here is unambiguous because getBearerToken below
+					// uses the same separator when formatting a remote request (and because
+					// repository names can't contain colons).
+					cacheKey = fmt.Sprintf("%s:%s:%s", extraScope.resourceType, extraScope.remoteName, extraScope.actions)
```

Comment:
I don't understand what you mean by that? We add the new scope as the parameter `makeRequestToResolvedURL` so that shouldn't require any adjustments.

Reply:
The concern described by this comment is that `cacheKey` should never map two different `extraScope` values into the same `cacheKey` value.

That could, looking _only_ at that code line, happen with `(resourceType = "a:b", remoteName = "c")`, vs. `resourceType = "a", remoteName = "b:c")`.

The comment above documents how that couldn’t happen in previous situations, where `actions` comes from our code and never contains colons, and `remoteName` is a repository name and thus doesn’t contain a colon.

Now that all of `extraScope` comes from an untrusted server, the same guarantees don’t trivially apply.

Maybe something like
```go
					// Using ':' as a separator here is unambiguous because getBearerToken below
 					// uses the same separator when formatting a remote request (and because
 					// repository names that we create can't contain colons, and extraScope values
 					// coming from a server come from `parseAuthScope`, which also splits on colons).
```

Reply:
Ah, that makes sense.

I have added a safety escape hatch here (better safe than sorry imho) which panics (In C/Python I'd include an `assert`), but if that's too brutal, then I can omit that or convert it to returning an error.

Reply:
Yes, please — `fmt.Errorf("internal inconsistency: …")` would be better here.

This is a “shouldn’t happen” situation, but if we are adding code to handle it, the added code shouldn’t increase the risk of this becoming a remote DoS of a long-running process.